collect_app/src/main/java/org/odk/collect/android/backgroundwork/ReentrantLockChangeLock.java

```diff
@@ -0,0 +1,18 @@
+package org.odk.collect.android.backgroundwork;
+
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.Function;
+
+public class ReentrantLockChangeLock implements ChangeLock {
+
+    private final ReentrantLock lock = new ReentrantLock();
+
+    @Override
+    public <T> T withLock(Function<Boolean, T> function) {
+        try {
+            return function.apply(lock.tryLock());
+        } finally {
+            lock.unlock();
```

Comment:
I can't imagine how it would happen but technically this could throw a `IllegalMonitorStateException`. Maybe catch and ignore out of an abundance of caution?

Reply:
Yeah I missed that in the docs. Thanks for pointing it out. Will add that in.